### PR TITLE
Acceptance test for --cli-version and --path options and updated help for install-cli

### LINKED ISSUE #326: Implement the --cli-version and --path options for install-cli
These options are not properly implemented yet. The help does not mention them, but they are mentioned in the README (going to remove it together with the Cucumber test which does not seem to do anything).

### PR DESCRIPTION
**Pull request by brgnepal**
Fix #326

### CODE REVIEW

**Comment by brgnepal** [comment]
@hferentschik , any idea how to test `--path` option?

Suppose, if I run command `vagrant service-manager install-cli docker --path <path>`

How do I validate this `path` in terms of current acceptance test framework?

**Comment by hferentschik** [comment]
@budhrg let me have a look into it, starting with CI ...

**Comment by hferentschik** [comment]
@budhrg, could you please push a rebased version of this pull request? This should also trigger the hopefully working CI job again.

**Comment by brgnepal** [comment]
@hferentschik It looks like rubocop check is not _really_ working locally

```
➜  vagrant-service-manager git:(fix-326) ✗ bundle exec rake rubocop
Running RuboCop...
Inspecting 47 files
...............................................

47 files inspected, no offenses detected
```

**Comment by brgnepal** [comment]
@hferentschik Pushing rebased version will not work. Some "guard related warning" in new changes I committed.
Will fix that and push along with rebase.

**Comment by brgnepal** [comment]
@hferentschik  Rubocop version was the culprit :wink: 

```
➜  vagrant-service-manager git:(fix-326) ✗ rubocop -v
0.42.0
➜  vagrant-service-manager git:(fix-326) ✗ rubocop lib/vagrant-service-manager/installer.rb
Inspecting 1 file
.

1 file inspected, no offenses detected
➜  vagrant-service-manager git:(fix-326) ✗ gem update rubocop
Updating installed gems
Updating rubocop
Fetching: rubocop-0.43.0.gem (100%)
Successfully installed rubocop-0.43.0
Parsing documentation for rubocop-0.43.0
Installing ri documentation for rubocop-0.43.0
Installing darkfish documentation for rubocop-0.43.0
Done installing documentation for rubocop after 12 seconds
Parsing documentation for rubocop-0.43.0
Done installing documentation for rubocop after 6 seconds
Gems updated: rubocop
➜  vagrant-service-manager git:(fix-326) ✗ rubocop lib/vagrant-service-manager/installer.rb
Inspecting 1 file
C

Offenses:

lib/vagrant-service-manager/installer.rb:39:9: C: Use a guard clause instead of wrapping the code inside a conditional expression.
        if @options.key?('--path') && !File.exist?(File.dirname(@options['--path']))
        ^^

1 file inspected, 1 offense detected
```

**Comment by brgnepal** [comment]
retest this please

**Comment by brgnepal** [comment]
retest this please

**Comment by brgnepal** [comment]
@hferentschik Build passing now.

I am unable to add the test for `--path` option. How you can put path and verify it in aruba test framework?

**Comment by hferentschik** [comment]
> I am unable to add the test for --path option. How you can put path and verify it in aruba test framework?

What exactly is the problem?

File: Gemfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ group :development do
   gem 'komenda', '~> 0.1.6'
   gem 'launchy'
   gem 'rake', '10.4.2'
-  gem 'rubocop'
+  gem 'rubocop', '~> 0.43.0'
```
Comment (hferentschik) [inline Gemfile]: Can you explain this change? How come that when I started from scratch as well, my Rubocop run worked, whereas on CI I got the errors? Are you saying different versions are getting installed depending on the environment?

**Comment by brgnepal** [comment]
> What exactly is the problem?

What path to specify in command when I say

```
When I run `bundle exec vagrant service-manager install-cli docker --path <which path>`
Then the exit status should be 0
And the binary "docker" should be installed with path "which path" (how do I verify here?)
```

File: lib/vagrant-service-manager/installer.rb [hunk 2]
```diff
@@ -5,15 +5,15 @@ def initialize(type, machine, env, options)
         @type = type
         @machine = machine
         @env = env
-        @box_version = options[:box_version]
+        @options = { type: @type }.merge(options)
```
Comment (hferentschik) [inline lib/vagrant-service-manager/installer.rb]: Is this not a bit odd? You pass _type_ and _option_ to this method, but then add _@type_ to the options as well? Would it not make more sense to add the type of the options before and just pass the options parameter?

Comment (hferentschik) [reply lib/vagrant-service-manager/installer.rb]: Alternatively, if _options_ is really just the options passed to the CLI and _type_ something you determine, keep passing both parameters along.

File: test/vagrant-service-manager/installer_test.rb [hunk 3]
```diff
@@ -29,7 +29,7 @@ module ServiceManager
           @installer.instance_variable_get('@type').must_equal @type
           @installer.instance_variable_get('@machine').must_equal @machine
           @installer.instance_variable_get('@env').must_equal @machine.env
-          @installer.instance_variable_get('@box_version').must_equal 'adb'
+          @installer.instance_variable_get('@options').must_equal({ type: @type }.merge(@options))
```
Comment (hferentschik) [inline test/vagrant-service-manager/installer_test.rb]: Even the test seems odd with this merge thing you are doing. Btw, I don't think these tests really do much. They are pretty much self fulfilling. I'd go so far as to say let's remove them

File: features/install-cli.feature [hunk 4]
```diff
@@ -52,6 +54,10 @@ Feature: Command behavior of client side tools installation
     Then the exit status should be 0
     And the binary "docker" should be installed
 
+    When I run `bundle exec vagrant service-manager install-cli docker --cli-version 1.12.1`
```
Comment (hferentschik) [inline features/install-cli.feature:57]: Any reason we only test the _docker_ download? Where are tests for downloading _oc_?

File: features/support/env.rb [hunk 5]
```diff
@@ -150,5 +150,15 @@ def extract_process_id(data)
   end
 
   expect(binaries.size).to eq(1)
-  expect(File.executable?(binaries[0])).to eq(true)
+  expect(File.executable?(binaries.first)).to eq(true)
+end
+
+Then(/^the binary "([^"]*)" should be installed with version "([^"]*)"$/) do |binary, version|
+  binaries = []
+  Find.find("#{ENV['VAGRANT_HOME']}/data/service-manager/bin") do |path|
```
Comment (hferentschik) [inline features/support/env.rb:158]: So here you are using the data directory to test something. Why can you not write a test which uses the --path options with a target of let's say  "#{ENV['VAGRANT_HOME']}/data/service-manager/custom-path". After completion the binary should be in the custom directory, right? BTW, if the custom directory does not exist, does it get created or what happens?

**Comment by hferentschik** [comment]
```
When I run `bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/data/service-manager/foo`
Then the exit status should be 0
And the binary "docker" should be installed with path "#{ENV['VAGRANT_HOME']}/data/service-manager/foo" (how do I verify here?)
```

Otherwise create tmpdir and let it install into the temp dir. IMO not so nice. Better to install into something which stays in the build directory.

File: Gemfile (hunk 1 above)
Comment (brgnepal) [reply Gemfile]: I don't know about the environment, but it seems version `0.42.0` not working for me at least please check your version.

More details here https://github.com/projectatomic/vagrant-service-manager/pull/391#issuecomment-250703761

Comment (brgnepal) [reply Gemfile]: Also, CI I think install latest always unless it is pinned down to some version @hferentschik

File: test/vagrant-service-manager/installer_test.rb (hunk 3 above)
Comment (brgnepal) [reply test/vagrant-service-manager/installer_test.rb]: @hferentschik, Ya agree. Just checking whether we are getting expected variables or not.

File: features/install-cli.feature (hunk 4 above)
Comment (brgnepal) [reply features/install-cli.feature:57]: @hferentschik I separated that with `adb/cdk-openshift.feature` to avoid duplication.

File: features/support/env.rb (hunk 5 above)
Comment (brgnepal) [reply features/support/env.rb:158]: @hferentschik If custom directory doesn't exist, if fails with invalid path message

Comment (brgnepal) [reply features/support/env.rb:158]: I don't know but `#{ENV['VAGRANT_HOME']}/data/service-manager/custom-path"` is not working in command

`bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/data/service-manager/docker`

File: Gemfile (hunk 1 above)
Comment (hferentschik) [reply Gemfile]: Hmm, I agree, rubocop version is somehow involved. I just don't understand. So looking at CU it indeed uses _0.43.0_ which is expected, since it is the latest release.

However, if I clean my gemset and execute a _bundle install_, I also should get this latest version, even w/o your change, right? But I get 0.41.2!?

File: test/vagrant-service-manager/installer_test.rb (hunk 3 above)
Comment (hferentschik) [reply test/vagrant-service-manager/installer_test.rb]: Usually, I would refrain from testing internal state or methods (there are exceptions of cause). Much better to test the external "contract" a class with its public method forms.

File: Gemfile (hunk 1 above)
Comment (hferentschik) [reply Gemfile]: ARGH, one of course has to delete the _Gemfile.lock_ as well. Otherwise even with deleting the gems, you only get the old versions back. Make sense now.

File: features/install-cli.feature (hunk 4 above)
Comment (hferentschik) [reply features/install-cli.feature:57]: but that's just the plain _install-cli_, right? I believe in this case we don't have the _--path_ and _--version_ option, right? Something we should test to verify the expected error message occurs. On the adb feature file we still can then test the 'oc' download with path and version.

File: features/support/env.rb (hunk 5 above)
Comment (hferentschik) [reply features/support/env.rb:158]: >  If custom directory doesn't exist, if fails with invalid path message

Which should be put in a test. Remember, these acceptance test also describe the expected behavior of all the commands, including the error cases.

Comment (hferentschik) [reply features/support/env.rb:158]: > I don't know but #{ENV['VAGRANT_HOME']}/data/service-manager/custom-path" is not working in command

What's the error? My guess is that the variable substitution does not work. Did you debug it? Check _aruba/cucumber/command.rb_ and _aruba/api/command.rb_. If backtick does not work, we can use _system_ and write our own step definition. Have a go at it.

**Comment by hferentschik** [comment]
@budhrg afaict nothing has changed since my last review, or am I missing something?

**Comment by brgnepal** [comment]
@hferentschik sorry couldn't update this PR as I am still finding the way to write test for --path option.

**Comment by hferentschik** [comment]
> sorry couldn't update this PR as I am still finding the way to write test for --path option.

no worries. For some reason I thought you wanted me to have another look.

**Comment by brgnepal** [comment]
@hferentschik Updated test for --path option. Used /tmp as of now. Not getting framework related path to use.

File: features/install-cli.feature [hunk 6]
```diff
@@ -52,6 +54,14 @@ Feature: Command behavior of client side tools installation
...
+    Then the exit status should be 0
+    And the binary "docker" should be installed with version "1.12.1"
+
+    When I run `bundle exec vagrant service-manager install-cli docker --path /tmp/docker`
```
Comment (hferentschik) [inline features/install-cli.feature]: I don't think that _/tmp/docker_ is a good choice. This won't work on Windows for example. I know we cannot run tests on Windows at, but we are not so far from it. This would just put another roadblock into our way. Also what would be responsible to clean this up, aka removing already downloaded binaries?

What exactly is the issue with using a env variable here? Where does it break? Do you have a stack trace or failing test output?

**Review by hferentschik — CHANGES_REQUESTED** [review]

File: .rubocop_todo.yml [hunk 7]
```diff
@@ -13,40 +13,33 @@ Lint/AmbiguousRegexpLiteral:
...
 
 # Offense count: 1
-Lint/HandleExceptions:
+Lint/Eval:
   Exclude:
-    - 'lib/vagrant-service-manager/service.rb'
+    - 'features/support/env.rb'
 
-# Offense count: 13
+# Offense count: 17
```
Comment (hferentschik) [inline .rubocop_todo.yml:20]: So we are getting worse? That's not good

File: .rubocop_todo.yml [hunk 8]
```diff
@@ -13,40 +13,33 @@ Lint/AmbiguousRegexpLiteral:
...
 Metrics/MethodLength:
   Max: 96
 
 # Offense count: 1
 # Configuration parameters: CountComments.
 Metrics/ModuleLength:
-  Max: 111
-  Exclude:
-    - 'lib/vagrant-service-manager/plugin_util.rb'
+  Max: 132
```
Comment (hferentschik) [inline .rubocop_todo.yml]: this will increase module length for all modules from now moving forward, opposed to exclude just _plugin_util.rb_, right?

File: .rubocop_todo.yml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # This configuration was generated by
 # `rubocop --auto-gen-config`
-# on 2016-07-09 23:34:13 +0200 using RuboCop version 0.41.2.
+# on 2016-10-13 17:03:27 +0530 using RuboCop version 0.43.0.
```
Comment (hferentschik) [inline .rubocop_todo.yml:3]: For what it's worth, I did not have to make any changes to the rubocop todo file. Is this change required?

File: features/install-cli.feature [hunk 10]
```diff
@@ -52,6 +54,15 @@ Feature: Command behavior of client side tools installation
     Then the exit status should be 0
     And the binary "docker" should be installed
 
+    When I run `bundle exec vagrant service-manager install-cli docker --cli-version 1.12.1`
+    Then the exit status should be 0
+    And the binary "docker" should be installed with version "1.12.1"
+
+    # '../vagrant.d' maps to ENV['VAGRANT_HOME']
+    When I run `bundle exec vagrant service-manager install-cli docker --path ../vagrant.d/docker`
```
Comment (hferentschik) [inline features/install-cli.feature:62]: So ENV['VAGRANT_HOME'] does not expand in here when used in the form of {#ENV['VAGRANT_HOME']...}?
Personally I would almost expect this to fail to begin with. Is it intentional to support relative paths?

File: locales/en.yml [hunk 11]
```diff
@@ -104,7 +104,9 @@ en:
             A supported service. For example: docker, kubernetes or openshift.
 
           Options:
-                -h, --help         print this help
+                -h, --help            print this help
+                --cli-version         binary version to install
+                --path                local path where to install the binary
```
Comment (hferentschik) [inline locales/en.yml]: should probably mention whether this path is absolute or relative or both

File: test/vagrant-service-manager/installer_test.rb (hunk 3 above)
Comment (hferentschik) [reply test/vagrant-service-manager/installer_test.rb]: @budhrg did you do something about this now? Remove?

**Review by hferentschik — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (hferentschik) [reply Gemfile]: Actually, I think this is technically not required. If you empty your gems and remove the Gemfile.lock all is good. You will automatically get 0.43. It is a bit sneaky to add this here.

File: .rubocop_todo.yml (hunk 8 above)
Comment (brgnepal) [reply .rubocop_todo.yml]: ya correct @hferentschik . Reverting it now.

File: features/install-cli.feature (hunk 10 above)
Comment (brgnepal) [reply features/install-cli.feature:62]: @hferentschik In this line, it is actually running as shell command, so `#{ENV['VAGRANT_HOME]}` form will not work here.

File: locales/en.yml (hunk 11 above)
Comment (brgnepal) [reply locales/en.yml]: Changed to "absolute or relative path where to install the binary"

File: .rubocop_todo.yml (hunk 9 above)
Comment (brgnepal) [reply .rubocop_todo.yml:3]: @hferentschik , I ran the command rubocop --auto-gen-config which gave me new set of offenses. I think this will help to monitor the list of offenses we are doing. 
For example, with new auto-gen-config, we found new offenses like below Abc count increased to 17 from 13.

File: .rubocop_todo.yml (hunk 7 above)
Comment (brgnepal) [reply .rubocop_todo.yml:20]: @hferentschik I think this offenses has been accumulated from last we ran `rubocop --auto-gen-config`. I think we need to create an issue to do a first set of rubocop offense cleanup.

File: Gemfile (hunk 1 above)
Comment (brgnepal) [reply Gemfile]: @hferentschik , Do you mean everytime, I need to remove Gemfile.lock file to map to ci like bundler environment?

File: features/install-cli.feature (hunk 4 above)
Comment (brgnepal) [reply features/install-cli.feature:57]: @hferentschik Add testcase for oc in adb-openshift.feature file

**Comment by brgnepal** [comment]
retest this please

File: .rubocop_todo.yml (hunk 7 above)
Comment (hferentschik) [reply .rubocop_todo.yml:20]: Might be worse to do all this rubocop fiddling separately. It is not really related to the issue at hand at all.

File: features/install-cli.feature (hunk 10 above)
Comment (hferentschik) [reply features/install-cli.feature:62]: well, usually backtick operator expands variables, right? Is this not the case here?

**Review by hferentschik — CHANGES_REQUESTED** [review]

File: lib/vagrant-service-manager/plugin_util.rb [hunk 12]
```diff
@@ -162,9 +162,11 @@ def self.fetch_existing_oc_binary_path_in_windows
         path_string = ENV['PATH']
 
         separator = ':' unless path_string.include?(';')
-        path_string.split(separator).detect do |dir_path|
+        oc_path_dir = path_string.split(separator).detect do |dir_path|
```
Comment (hferentschik) [inline lib/vagrant-service-manager/plugin_util.rb:165]: So what is this change about? How is it related to the issue? This is a Windows issue and not related to this added tests, right? Is there another issue on Windows which gets fixed here?

Comment (brgnepal) [reply lib/vagrant-service-manager/plugin_util.rb:165]: @hferentschik Yes, I forgot to return the detected oc.exe path. Instead I was returning the first detected `dir_path`

**Comment by brgnepal** [comment]
@hferentschik Also added documentation in `commands.adoc`

File: commands.adoc [hunk 13]
```diff
@@ -223,10 +223,18 @@ Similarly, you can use `$ vagrant service-manager status openshift`,
 
 If you need to install the client binary for a specified service, use: +
 
-`vagrant service-manager install-cli [service]` +
+`vagrant service-manager install-cli [service] [options] [arg]` +
 
 The possible options for service are `docker` and `openshift`. +
 
+The following options can be used: +
```
Comment (brgnepal) [inline commands.adoc:230]: @Preeticp Please have a quick look to this docs.

cc @hferentschik

File: commands.adoc [hunk 14]
```diff
@@ -223,10 +223,18 @@ Similarly, you can use `$ vagrant service-manager status openshift`,
...
+
+`--cli-version`::  Version of binary to install +
+`--path`::   Absolute or relative path where to install the binary +
```
Comment (Preeticp) [inline commands.adoc:233]: Replace:
Version of binary to install>Displays the version of the binary to be installed.
Absolute or relative path where to install the binary>Displays the absolute or relative path where the binary should be installed.

File: commands.adoc [hunk 15]
```diff
@@ -223,10 +223,18 @@ Similarly, you can use `$ vagrant service-manager status openshift`,
...
+`-h, --help`::   Prints help for the specific command being run.
+
+_arg_ represents the valid value for the respective `--cli-version` and/or `--path` options.
+
```
Comment (Preeticp) [inline commands.adoc:237]: This is the only place we are using Italics for words like '_arg_'. I would prefer we use `[arg]`/`arg` to keep it consistent with the style and also to prevent confusion for users. However, whichever style we decide to use, it would have to be applied to the entire document for words like, commands, sub-commands, options etc. Or we can simply change _arg_ to arg.

Comment (Preeticp) [inline commands.adoc:237]: Also, should we modify the example below to show the --cli version and --path options you have added?

File: commands.adoc (hunk 14 above)
Comment (hferentschik) [reply commands.adoc:233]: @Preeticp "Displays"!? This commands let you specify a version and a path, they are not displaying them

**Comment by hferentschik** [comment]
retest

**Comment by hferentschik** [comment]
Seems that we have Rubocop failures again :-( This time a Rubocop 0.44 got releases. I guess we need to thing about, whether we want to pin the Rubocop version to prevent this to happen or whether we just handle it as it comes. Either way, let's do this separately. Let's create an issue and pull request for upgrading (and potentially pinning version) of Rubocop. Once this is in place we can place this pull request on top. @budhrg Any strong feeling regarding setting a fixed version number vs dealing with new "violations"?

**Comment by hferentschik** [comment]
@budhrg I am also closing this pull request. Let's split Rubocop from install-cli and create new pull requests. We take it from there.

File: commands.adoc (hunk 14 above)
Comment (brgnepal) [reply commands.adoc:233]: Ya @Preeticp , Probably should use "Specify version of binary to install" and "Specify absolute or relative path where to install .."

Comment (Preeticp) [reply commands.adoc:233]: Yes, Specify the version of the binary to be installed. and Specify the absolute or relative path where the binary should be installed.

**Comment by brgnepal** [comment]
> feeling regarding setting a fixed version number 

@hferentschik 

I am in favor of pinning to fix version and update it later on instead of dealing with new violations every time new version is released.

**Comment by brgnepal** [comment]
Also, will give the PR of this issue https://github.com/projectatomic/vagrant-service-manager/issues/412 first @hferentschik
